index.bs

```diff
@@ -543,54 +544,85 @@ To <dfn>handle an incoming message</dfn> given a [=WebSocket connection=]
-    throws an exception, then [=respond with an error=] given
-    |connection| and [=error code=] [=invalid argument=], and finally
-    return.
-    <!-- corresponds to Parse error (-32700) in JSON-RPC -->
-
- 5. Let |result| be the result of [=process a command|processing a
+ 1. Let |result| be the result of [=process a command|processing a
     command=] given |data|.
 
- 6. If |result| is an [=Error=], then [=respond with an error=] given
-    |connection|, |result|, and |parsed|["<code>id</code>"], and finally
-    return.
+ 1. If |result| is an [=error=], then [=respond with an error=] given
+    |connection|, |data|, and |result|'s [=error code=], and finally return.
+
+ 1. Assert: |result| is a [=success=]
 
- 7. Let |response| be a new [=map=] with the following properties:
+ 1. Let |command id| and |value| be the two fields |result|'s data.
+
+ 1. Let |response| be a new [=map=] with the following properties:
 
    <dl>
       <dt>"id"</dt>
-      <dd>The value of |parsed|["<code>id</code>"]</dd>
+      <dd>|command id|</dd>
 
-      <dt>"result"</dt>
-      <dd>The value of |result|</dd>
+      <dt>"value"</dt>
+      <dd>|value|</dd>
    </dl>
 
- 8. Let |serialized| be the result of [=serialize JSON to
-    bytes|serializing JSON to bytes=] given |response|.
+ 1. Let |serialized| be the result of [=serialize a map as JSON=] given
+    |response|.
 
- 9. [=Send a WebSocket message=] comprised of |serialized| over
+ 1. [=Send a WebSocket message=] comprised of |serialized| over
     |connection|.
 
 </div>
 
 <div algorithm>
 To <dfn>respond with an error</dfn> given a [=WebSocket connection=]
-|connection| and an [=error code=] |code|:
+|connection|, |data| and |error code|:
 
- 1. Issue: Form a valid JSON |errorObject| given |code|.
+ 1. Let |command id| be null.
 
- 2. [=Send a WebSocket message=] comprised of |errorObject| over
-    |connection|.
+ 1.  Let |parsed| be the result of [=parse JSON into Infra values|parsing JSON
+     into Infra values=] given |data|. If this does not throw an exception, and
+     results in a map that contains "<code>id</code>", then let |command id| be
+     |parsed|["<code>id</code>"].
+
+ 1. If |command id| is not an integer greater than or equal to zero, let
+    |command id| be null.
+
+ 1. Let |error data| be a new map matching the <code>ErrorResponse</code>
+    production in the [=local end definition=], with the <code>id</code> field
+    set to |command id|, the <code>error</code> field set to |error code|, the
+    <code>message</code> field set to an implementation-defined string
+    containing a human-readable definition of the error that occured and the
+    <code>stacktrace</code> field optionally set to an implementation-defined
+    string containing a stack trace report of the active stack frames at the
+    time when the error occurred.
+
+ 1. Let |response| be the result of [=serialize a map as JSON=] given |error
```

Comment:
We can now use https://infra.spec.whatwg.org/#serialize-an-infra-value-to-a-json-string here.

Reply:
Presumably we want to use "as bytes" instead.

Reply:
Yes, my bad.